Modernization, Sticker Support, Advanced `find` functions...

m7a opened the pull request:
Hello,

while there is no public code for my client yet (it's a slow work-in-progress), I figured I could share some of the patches that I have created for the `erlmpd` library. It is currently an assorted set of changes that I found useful while working on  my client.

I don't mind maintaining an own fork, but in case some of the patches also look useful to you, feel free to include them into the upstream or provide comments. In case a certain subset looks interesting to you, I can also split up this PR as needed.

The patches address the following topics:

 * The first commits reduce warnings and enable rebar3 support to make including the library as a dependency easier
 * I enabled the sticker code and worked on parsing some of the sticker command's outputs. In the most recent changes I also gave the sticker operations dedicated functions (this makes declaring the types properly easier and since the code had been commented-out before I figured the API needed not to be stable yet :smile:)
 * find() newly supports an advanced query syntax. I included a patch to build this syntax with all the correct quoting and aligned to the MPD protocol specification.
 * I had some fun with idle/noidle interactions and proposed a preliminary workaround for the issue. It may require some additional patch to resolve it completely, but I couldn't figure out how to do it completely just yet hence I suggest a one-line workaround :smile:

Thanks in advance and Kind regards
Linux-Fan (@m7a)

clonejo committed:
fix build script (edoc)

m7a committed:
Enable stickers support (comment-in the previously unused code). It works just fine in my tests.

m7a committed:
Enable compilation through rebar3 build tool.

m7a committed:
erlmpd.erl: Remove definition of binary_to_atom newly provided by the Erlang runtime itself.

m7a committed:
Change from @spec to -spec to fix warnings generated by EDoc and make use of the type info as far as it is available.

m7a committed:
Add disconnect() call because this should newly be favored over close() command according to MPD documentation. Also, it seems that previously, the connection lifecycle could not be managed explicitly. This commit fixes this by providing the new disconnect() function.

m7a committed:
Add find() call with a single parameter to allow using the “new” syntax where a query can be sent to MPD. Currently, this query needs to be assembled by the using application. It may make sense to provide a means to build the respective expressions using the library, too, later.

m7a committed:
Align return code of sticker set operation with remainder of operations.

m7a committed:
escape_quotes: Correct escaping in the presence of nested quotes

In most cases, nested quotes are not needed. When dealing with expressions
passed to find, they can appear, though. In these cases it becomes relevant to
escape backslashes that may already be present in the input string in addition
to handling the regular string escapes.

This commit changes the escape_quotes function accordingly.

m7a committed:
Add auxiliary functions for building find() expressions

Previously, queries passed to find() had to be provided as correctly-quoted
strings directly to the function. Since this way, not much of the abstraction
features of such a library are leveraged, this commit replaces the previously
created string-based API with one that takes a tuple that represents
FILTER expression constructed of Erlang terms.

Usage could e.g. look as follows:

	erlmpd:find(Conn, {and, [{lnot, {tagop, artist, eq, "val"}},
	                         {tagop, title,  eq, "val2"}]}).

m7a committed:
Revise sticker handling

Previously, unprocessed data was returned by most of the sticker functions.
This would require every application to handle this part of the MPD protocol
for itself.

This commit revises the sticker API to become more user-friendly:

 * It splits out all of the different functions (find/delete/get/...) to
   dedicated functions. This makes documenting the individual routines
   easier and also allows clear type specifications to be formulated.

 * It adjusts all the “reading” functions to parse the sticker values and
   discard redundant info (such as but not limited to the leading "sticker: "
   prefix etc). As a result, applications can now directly process the values
   they are interested in.

Note that currently, the find function is the only one to return binaries
wherehas the other ones work on strings. It is not exactly clear why this
happens, though. It may turn out that in the end the return value is better
processed as iolist?

m7a committed:
idle: Pass errors returned from command

Previously, errors returned by the `idle` command were ignored.
Occasionally, when interacting with `noidle` and `idle`, errors of type
`ealready` were observed.

The correct solution in these cases is to ensure a proper alternation of
`idle` and `noidle` commands on the TCP level. This requires a means for idle()
to report when it has completed the `gen_tcp:send` of the CommandStr in order
for an application to only call `command("noidle")` after having received this
confirmation. This is complicated to implement on the library side and not much
easier on the application side.

As a workaround for the issue, this commit changes the `idle()` function to
return the error. It was checked that applications can handle the error and
continue to use the connection afterwards. A proper solution for this issue
could be implemented by a later commit.

The error handling provided by this commit should be fine independently of
future changes to the idle handling to fix the described `noidle`interaction.

m7a committed:
count/2, count_group/3: Add new count APIs

Previously, an interface to MPD's count API was only provided by count/3 with
the intention of querying the count for exactly one tag match. It could already
be used for very simple “group” queries by telling the API that the tag of
interest was `group`. However, this resulted in the output list being difficult
to parse for the application and filtering could not be used in conjunction with
this API.

This commit attempts to solve these limitations as follows:

 * It adds a new count/2 API that is intended to produce an output format equal
   to the count/3 call but has enhanced querying capabilities by allowing
   arbitrary filter expressions to be used in place of a single tag match.

 * It adds a new count_group API that can be used to query multiple counts at
   once by grouping after a certain tag. It parses the resulting output inside
   the erlmpd library and outputs a list of property lists similar to other
   (e.g. song-based) APIs.

caolan commented:
@m7a wow, it's great you're continuing to work on and modernize this library. As I'm unlikely to find time to review your changes, perhaps the most sensible thing to do is for you to fork the repository and me to point to your fork in the README?

m7a commented:
> As I'm unlikely to find time to review your changes, perhaps the most sensible thing to do is for you to fork the repository and me to point to your fork in the README?

This is indeed a good option. For now I am going to continue working on my patches alongside the development of my player application. Once the player is ready, I'd appreciate a link from the README.

In the meantime it might make sense to keep it just in this state such that interested users can find the code by looking at the pull request but are not in any way encouraged to assume that my version is going to be well-maintained/canonical just yet :smile:

Thank you for this project btw. the code has been very helpful in getting started with MPD client development and many of my changes were on a “cosmetic” level only so far (except for the filter stuff maybe...).

caolan commented:
Sure, feel free to keep adding code here if you prefer.

m7a committed:
Use proper HTML entities in documentation commits. This fixes an error message being printed during the generation of the documentation.

m7a committed:
idle: Provide low-level functions to ensure a safe interaction between idle and noidle

In 3df96857d1 ("idle: Pass errors returned from command"), a workaround for
errors thrown by idle/1 and idle/2 was proposed. It allowed these errors to be
recognized by an application, however, the underlying issue had not been
addressed yet.

The issue is that noidle is only valid to send after an idle command was sent
right before. Since both operations are supposed to behave asynchronously, this
could not be established by applications.

This commit provides the building blocks that application developers can use to
implement a solution for this issue e.g. by using erlang processes. The idea
is as follows:

 * Strictly speaking it is impossible to assure the order of requests in this
   case.
 * However, even if the requests arrive in a chaotic order, a “gating” process
   could ensure that the order of commands delivered is always a valid
   alternation of idle and noidle e.g. by dropping noidle calls or inserting
   additional “dummy” idle calls into the stream
 * To do this, such a process must distinguish between the two stages of the
   idle command: First stage is to send idle to MPD and second stage is to
   wait for a response. For the proper ordering of commands it is vital that
   the process can make a valid assumption about when the first stage was
   performed because even without having started the second stage yet it is
   valid to call noidle afterwards.

To provide applications with the necessary interfaces, a split version of the
idle-API is provided called idle_send/2 and idle_receive/1. After idle_send/2
has returned successfully, noidle/1 may be called even if idle_receive/1 has not
been started yet.

Since it is now possible to use this command stably with the API, noidle/1 has
been given its own dedicated API call (command(C, "noidle") was possible before
already...)

Despite these new options it remains a viable solution for application
developers to do aside with the asynchronous `noidle` and prefer to open two
connections (one for idle purposes and one for actions). Such a design may be
more robust than trying to get idle/noidle interact correctly...

m7a committed:
addid_relative: New function to add relative to currently playing song

Previously, addid supported two modes of operation:

 * Adding at an absolute position in the playlist
   (with non-negative integer parameter)
 * Adding at a relative position before the current song
   (with negative integer parameter)

This commit enables the following additional option:

 * Adding at a relative position after the current song

To do this, a new function `addid_relative` is provided that always
(independent of its sign) interprets the `Pos` parameter as a relative
offset in the playlist.

m7a committed:
Plalists: Remove redundant `pass_errors` invocations

Previously, API endpoints plchangesposid/2 and listplaylists/1 used dedicated
subfunctions to conditionally execute parse_group/2 only in the absence of
errors by using pass_errors/2. The implementation of parse_group/2 contains
another invocation of pass_errors/2, though. Hence the outer pass_errors call
is redundant and can be removed.

This commit removes the superflous pass_errors calls and inlines the newly
trivial subfunctions.

m7a committed:
Partitions: Expose partitions features through new API calls

Previously, partition management was not explicitly exposed through erlmpd.
While it could be invoked through the generic `command` API this moved the
burden of properly parsing the protocol results to the client.

This commit adds new dedicated API calls for partition management. The names
are equal to the ones used in the MPD protocol specification.

m7a committed:
partition: Fix documentation comment to avoid errors upon documentation generation.

m7a committed:
Documentation generation: Include CSS and PNG resources in the documentation output. This improves the rendering quality of the generated documentation and avoids leaving stray files in the src directory after generating documentation.

m7a committed:
noidle: Fix {error,ealready} and timeouts

Previously, `noidle` was handled like any other command.
Specifically, a response was expected to be provided by MPD and intended to be
processed by `gen_tcp:recv`. However, in typical `noidle` use cases, a
`gen_tcp:recv` may already be ongoing (could even be from inside another
process).

In these cases it is harmful to expect a response because it can lead to
important response contents being dropped and also often fails with
`ealready` errors right away.

This commit fixes this by sending `noidle` commands without reading a
response from MPD afterwards.

m7a committed:
Filter: Fix generation of wrong queries

Previously, when using filters like `{tagop, title, contains, Query}` this
would generate something like `titlecontains "Query"` and send it to MPD. This
is of course incorrect and spacing is necessary between the operands.

This commit adds spaces around the operators as to assure that the generated
queries are correct.

m7a committed:
Music database: Add search/2 API

Previously, there was only the find/2 API when intending to use filter
expressions. This commit adds an analogous serach/2 API. They differ in
that search is case-insensitive by default making it well-suited for
queries based on user input.

m7a committed:
Add toggleoutput/1 API. This is in alignment with the existent disableoutput and enableoutput routines and helps in cases where the application logic may only display a state in the GUI but not actually track the state of the outputs.

m7a committed:
sticker_find/7: Provide API for sticker searching with filter

Previously, sticker searching was only possible by name.
This commit adds support for the MPD protocol:

	sticker find {TPYE} {URI} {NAME} = {VALUE} [sort {SORTTYPE}] [window {START:END}]

The new functionality is provided by sticker_find/7. Although it is the first
function to allow making use of MPD's output sorting and pagination
capabilities, these have been implemented in a manner that allows extending the
same mechanism to other commands as needed, too.

The idea is to have an `Options` last argument that may take a proplist with
optional arguments including those for `window` and `sort`. Note that the
valid values to pass for `sort` differ between the standard `find` and
`sticker find`.

m7a commented:
I have extended my fork into a fully-fledged copy including links to documentation. Find the description here: <https://masysma.net/32/erlmpd.xhtml>.

Feel free to add a link from the README in case you consider the fork a worthwhile continuation of the original idea behind the library.

Once I figure out some details, I am also going to put online a first alpha of my own client. I shall keep you posted about this.

m7a commented:
My own client based on erlmpd + my patches is finally there:
<https://masysma.net/32/maenmpc.xhtml>.

m7a committed:
albumart, readpicture: Add new commands

Previously, erlmpd didn't support parsing binary responses from MPD. These kinds
of responses are defined to be returned by fucntions related to reading image
data such as the `albumart` and `readpicture` APIs.

This commit adds support for commands which may return binary data and makes use
of this to implement albumart/2 and readpicture/2. Unlike the commands on the
MPD level, these commands block until all of the data was returned. This makes
application-side processing easier at the expense of losing some flexibility
wrt. downloadnig image contents in the background.

The suggested implementation for users that need to do this in the background is
to offload the downloading of the image data to a separate process and
connection.

caolan commented:
I've added a link to your fork in the README.

m7a committed:
find_ex/3: Add new command

Previously, sort and window options had not been supported for find.
This commit adds this functionality. As find/3 is already assigned to a
different semantics for backwards compatibility, a new identifier find_ex/3
was chosen.

m7a committed:
sort_option: Fix regression in docs generation ` -> '

m7a committed:
count_group: Improve documentation formatting.

m7a committed:
find_ex, tagtypes: Add example output/invocation to documentation.

m7a committed:
tagtypes_disable, tagtypes_enable, tagtypes_clear, tagtypes_all: Extend tagtypes series of commands

Previously, these commands were not exposed through the API. Since they help
with performance optimization of clients, this patch adds support for the
API functions:

 * tagtypes_disable
 * tagtypes_enable
 * tagtypes_clear
 * tagtypes_all

Like for filters, the tag types may be supplied as atoms.

m7a committed:
window_option: Fix it such that the intended syntax actually works.

Before it was impossible to use the option because it would cause a runtime
exception in io_lib:format. Also, the required format of the property was
documented wrongly.

This commit fixes this.

m7a committed:
[BREAKING change] sticker delete: Return ok in success to align it with other erlmpd APIs.

m7a committed:
volume: Implement relative volume change command.

m7a committed:
sticker_inc, sticker_dec: Add new functions to expose these MPD API capabilities. Document the curious asymmetry between sticker_inc and sticker_dec in the description for tiscker_dec.

m7a committed:
sticker commands: Add support for type "filter".
